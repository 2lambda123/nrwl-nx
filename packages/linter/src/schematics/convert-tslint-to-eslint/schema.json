{
  "$schema": "http://json-schema.org/schema",
  "id": "convert-tslint-to-eslint",
  "title": "Convert an Nx Angular project from TSLint to ESLint",
  "description": "NOTE: Does not work in the --dry-run mode",
  "examples": [
    {
      "command": "g convert-tslint-to-eslint myapp",
      "description": "Convert the Nx Angular project `myapp` from TSLint to ESLint"
    }
  ],
  "type": "object",
  "properties": {
    "project": {
      "description": "The name of the project to convert.",
      "type": "string",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "Which project would you like to convert from TSLint to ESLint?"
    },
    "removeTSLintIfNoMoreTSLintTargets": {
      "type": "boolean",
      "description": "If this conversion leaves no more TSLint usage in the workspace, it will remove TSLint and related dependencies and configuration",
      "default": true,
      "x-prompt": "Would you like to remove TSLint and its related config if there are no TSLint targets remaining after this conversion?"
    }
  },
  "required": ["project"]
}
