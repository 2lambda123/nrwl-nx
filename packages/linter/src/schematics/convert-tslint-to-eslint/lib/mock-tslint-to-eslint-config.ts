import {
  exampleE2eProjectTslintJson,
  exampleAngularProjectTslintJson,
  exampleRootTslintJson,
  exampleNonAngularProjectTslintJson,
} from './example-tslint-configs';

/**
 * The actual `findReportedConfiguration()` function is used to execute
 * `tslint --print-config` in a child process and read from the real
 * file system. This won't work for us in tests where we are dealing
 * with a Tree, so we mock out the responses from `findReportedConfiguration()`
 * with previously captured result data from that same command.
 */

export function mockFindReportedConfiguration(_, pathToTSLintJson) {
  if (
    pathToTSLintJson === 'tslint.json' ||
    pathToTSLintJson === '/tslint.json'
  ) {
    return exampleRootTslintJson.tslintPrintConfigResult;
  }

  if (
    pathToTSLintJson === 'apps/app1/tslint.json' ||
    pathToTSLintJson === 'libs/lib1/tslint.json'
  ) {
    return exampleAngularProjectTslintJson.tslintPrintConfigResult;
  }

  if (pathToTSLintJson === 'apps/app1-e2e/tslint.json') {
    return exampleE2eProjectTslintJson.tslintPrintConfigResult;
  }

  if (pathToTSLintJson === 'apps/app2/tslint.json') {
    return exampleNonAngularProjectTslintJson.tslintPrintConfigResult;
  }

  throw new Error(
    `${pathToTSLintJson} is not a part of the supported mock data for these tests`
  );
}
