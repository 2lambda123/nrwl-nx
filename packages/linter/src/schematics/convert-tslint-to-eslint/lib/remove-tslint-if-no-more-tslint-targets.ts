import { chain, Rule } from '@angular-devkit/schematics';
import { NodePackageInstallTask } from '@angular-devkit/schematics/tasks';
import {
  readWorkspace,
  updateJsonInTree,
  updateWorkspaceInTree,
} from '@nrwl/workspace';

/**
 * If no more instances of @angular-devkit/build-angular:tslint
 * are found within the user's workspace, we can remove the relevant
 * TSLint dependencies and any applicable global config.
 */
export function removeTSLintIfNoMoreTSLintTargets(
  rootTSLintJsonPath: string
): Rule {
  return (tree, context) => {
    const workspaceJSON = readWorkspace(tree);

    for (const projectName of Object.keys(workspaceJSON.projects)) {
      const targets =
        workspaceJSON.projects[projectName].targets ||
        workspaceJSON.projects[projectName].architect;

      for (const [_, targetConfig] of Object.entries(targets)) {
        const executor =
          (targetConfig as any).executor || (targetConfig as any).builder;
        if (executor === '@angular-devkit/build-angular:tslint') {
          // Workspace is still using TSLint, exit early
          return;
        }
      }
    }

    // If we got this far the user has no remaining TSLint usage

    tree.delete(rootTSLintJsonPath);

    const depsToRemove = ['tslint', 'codelyzer'];

    return chain([
      updateJsonInTree('package.json', (json) => {
        let hasChanges = false;
        depsToRemove.forEach((depName) => {
          if (json.dependencies?.[depName]) {
            delete json.dependencies[depName];
            hasChanges = true;
          }
          if (json.devDependencies?.[depName]) {
            delete json.devDependencies[depName];
            hasChanges = true;
          }
        });
        if (hasChanges) {
          context.addTask(new NodePackageInstallTask());
        }
        return json;
      }),
      updateWorkspaceInTree((json) => {
        json.schematics = json.schematics || {};

        /**
         * At the time of creating this schematic, ESLint is not the default for Angular
         * projects, and the "linter" property is not configured in the JSON, so we need
         * to explicitly make sure the relevant config is included.
         */
        json.schematics['@nrwl/angular:application'] =
          json.schematics['@nrwl/angular:application'] || {};
        json.schematics['@nrwl/angular:application'].linter = 'eslint';

        json.schematics['@nrwl/angular:library'] =
          json.schematics['@nrwl/angular:library'] || {};
        json.schematics['@nrwl/angular:library'].linter = 'eslint';

        function isJSONObjectLiteral(o: unknown) {
          return Boolean(o && o.constructor === Object);
        }

        function recursivelyCheckForAndUpdateLinter(configProp, config) {
          const maybeNestedConfigObj = config[configProp];

          if (isJSONObjectLiteral(maybeNestedConfigObj)) {
            for (const nestedConfigProp of Object.keys(maybeNestedConfigObj)) {
              recursivelyCheckForAndUpdateLinter(
                nestedConfigProp,
                maybeNestedConfigObj
              );
            }
          }

          if (configProp !== 'linter') {
            return;
          }
          config[configProp] = 'eslint';
        }

        for (const [, config] of Object.entries(json.schematics)) {
          for (const configProp of Object.keys(config)) {
            recursivelyCheckForAndUpdateLinter(configProp, config);
          }
        }

        return json;
      }),
    ]);
  };
}
