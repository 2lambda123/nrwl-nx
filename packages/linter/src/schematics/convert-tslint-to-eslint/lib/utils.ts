import { join, normalize } from '@angular-devkit/core';
import { Rule, SchematicContext, Tree } from '@angular-devkit/schematics';
import { addDepsToPackageJson } from '@nrwl/workspace';
import { allFilesInDirInHost } from '@nrwl/workspace/src/utils/ast-utils';
import * as assert from 'assert';

export function updateArrPropAndRemoveDuplication(
  json: Record<string, any>,
  configBeingExtended: Record<string, any>,
  arrPropName: string,
  deleteIfUltimatelyEmpty: boolean
): void {
  json[arrPropName] = json[arrPropName] || [];
  configBeingExtended[arrPropName] = configBeingExtended[arrPropName] || [];
  json[arrPropName] = json[arrPropName].filter(
    (extended: string) => !configBeingExtended[arrPropName].includes(extended)
  );
  json[arrPropName] = Array.from(new Set(json[arrPropName]));
  if (deleteIfUltimatelyEmpty && json[arrPropName].length === 0) {
    delete json[arrPropName];
  }
}

export function updateObjPropAndRemoveDuplication(
  json: Record<string, any>,
  configBeingExtended: Record<string, any>,
  objPropName: string,
  deleteIfUltimatelyEmpty: boolean
): void {
  json[objPropName] = json[objPropName] || {};
  configBeingExtended[objPropName] = configBeingExtended[objPropName] || {};

  for (const [name, val] of Object.entries(json[objPropName])) {
    const valueOfSamePropInExtendedConfig =
      configBeingExtended[objPropName][name];

    try {
      assert.deepStrictEqual(val, valueOfSamePropInExtendedConfig);
      delete json[objPropName][name];
    } catch {}
  }

  if (deleteIfUltimatelyEmpty && Object.keys(json[objPropName]).length === 0) {
    delete json[objPropName];
  }
}

export function ensureESLintPluginsAreInstalled(
  eslintPluginsToBeInstalled: string[]
): Rule {
  return (_host: Tree, context: SchematicContext) => {
    if (!eslintPluginsToBeInstalled?.length) {
      return;
    }

    const additionalDevDependencies = {};

    for (const pluginName of eslintPluginsToBeInstalled) {
      additionalDevDependencies[pluginName] = 'latest';
    }

    context.logger.info(
      '\nINFO: To most closely match your tslint.json, we will ensure the `latest` version of the following eslint plugin(s) are installed:'
    );
    context.logger.info('\n  - ' + eslintPluginsToBeInstalled.join('\n  - '));
    context.logger.info(
      '\nPlease note, you may later wish to pin these to a specific version number in your package.json, rather than leaving it open to `latest`.\n'
    );

    return addDepsToPackageJson({}, additionalDevDependencies);
  };
}

/**
 * Reliably inferring that something is an Angular project is not super straightforward,
 * particularly in the case of libraries, and we have to try a number of different heuristics:
 *
 * - is it configured to use a builder from @angular-devkit (other than the TSLint one itself of course)
 * - it has a karma.conf.js at the root of the project
 * - it has a jest.config.js at the root of the project which contains "jest-preset-angular"
 *
 * NOTE: we cannot do something like "contains a file ending in .module.ts", because NestJS has used so
 * many of the Angular conventions in its file naming.
 */
export function isAngularProject(tree: Tree, projectConfig: any): boolean {
  /**
   * @angular-devkit target heuristic
   */
  const targets = projectConfig.targets || projectConfig.architect;
  for (const [, targetConfig] of Object.entries(targets)) {
    const executor =
      (targetConfig as any).executor || (targetConfig as any).builder;
    if (
      executor &&
      executor.startsWith('@angular-devkit/') &&
      executor !== '@angular-devkit/build-angular:tslint'
    ) {
      return true;
    }
  }
  /**
   * Unit test configuration heuristic
   */
  if (tree.exists(join(normalize(projectConfig.root), 'karma.conf.js'))) {
    return true;
  }
  if (tree.exists(join(normalize(projectConfig.root), 'jest.config.js'))) {
    if (
      tree
        .read(join(normalize(projectConfig.root), 'jest.config.js').toString())
        .includes('jest-preset-angular')
    ) {
      return true;
    }
  }
  return false;
}

export function isE2EProject(projectName: string, projectConfig: any): boolean {
  const targets = projectConfig.targets || projectConfig.architect;
  return projectName.endsWith('-e2e') || !!targets.e2e;
}
