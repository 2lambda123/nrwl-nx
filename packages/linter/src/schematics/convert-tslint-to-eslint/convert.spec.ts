import { join, normalize } from '@angular-devkit/core';
import { chain, Tree } from '@angular-devkit/schematics';
import {
  readWorkspace,
  updateJsonInTree,
  updateWorkspace,
} from '@nrwl/workspace';
import {
  allFilesInDirInHost,
  readJsonInTree,
} from '@nrwl/workspace/src/utils/ast-utils';
import { callRule, createEmptyWorkspace } from '@nrwl/workspace/testing';
import { runSchematic } from '../../utils/testing';
import {
  exampleE2eProjectTslintJson,
  exampleAngularProjectTslintJson,
  exampleRootTslintJson,
  exampleNonAngularProjectTslintJson,
} from './lib/example-tslint-configs';
import { mockFindReportedConfiguration } from './lib/mock-tslint-to-eslint-config';
import { ConvertTSLintToESLintSchema } from './schema';

/**
 * See ./mock-tslint-to-eslint-config.ts for why this is needed
 */
jest.mock('tslint-to-eslint-config', () => {
  return {
    ...jest.requireActual('tslint-to-eslint-config'),
    findReportedConfiguration: jest
      .fn()
      .mockImplementation(mockFindReportedConfiguration),
  };
});

jest.mock('child_process', () => {
  return {
    ...jest.requireActual('child_process'),
    execSync: jest.fn().mockImplementation(() => {}),
  };
});
jest.mock('fs', () => {
  return {
    ...jest.requireActual('fs'),
    writeFileSync: jest.fn().mockImplementation(() => {}),
  };
});

jest.mock('@nrwl/eslint-plugin-nx', () => {
  return {
    configs: {
      angular: {
        env: {
          browser: true,
          es6: true,
          node: true,
        },
      },
      'angular-template': {},
    },
  };
});

describe('convert-tslint-to-eslint', () => {
  let tree: Tree;

  beforeEach(async () => {
    tree = Tree.empty();
    tree = createEmptyWorkspace(tree);
    tree = await callRule(
      updateWorkspace((workspace) => {
        workspace.projects.add({
          name: 'app1',
          root: 'apps/app1',
          sourceRoot: 'apps/app1/src',
          projectType: 'application',
          architect: {
            serve: {
              builder: '@angular-devkit/build-angular:dev-server',
              options: {
                browserTarget: 'app1:build',
              },
              configurations: {
                production: {
                  browserTarget: 'app1:build:production',
                },
              },
            },
            lint: {
              builder: '@angular-devkit/build-angular:tslint',
              options: {
                tsConfig: [
                  'apps/app1/tsconfig.app.json',
                  'apps/app1/tsconfig.spec.json',
                  'apps/app1/tsconfig.editor.json',
                ],
                exclude: ['**/node_modules/**', '!apps/app1/**/*'],
              },
            },
          },
        });
        workspace.projects.add({
          name: 'app1-e2e',
          root: 'apps/app1-e2e',
          sourceRoot: 'apps/app1-e2e/src',
          projectType: 'application',
          architect: {
            e2e: {
              builder: '@nrwl/cypress:cypress',
              options: {
                cypressConfig: 'apps/app1-e2e/cypress.json',
                tsConfig: 'apps/app1-e2e/tsconfig.e2e.json',
                devServerTarget: 'app1:serve',
              },
              configurations: {
                production: {
                  devServerTarget: 'app1:serve:production',
                },
              },
            },
            lint: {
              builder: '@angular-devkit/build-angular:tslint',
              options: {
                tsConfig: ['apps/app1-e2e/tsconfig.e2e.json'],
                exclude: ['**/node_modules/**', '!apps/app1-e2e/**/*'],
              },
            },
          },
        });
        workspace.projects.add({
          name: 'lib1',
          root: 'libs/lib1',
          sourceRoot: 'libs/lib1/src',
          projectType: 'library',
          architect: {
            lint: {
              builder: '@angular-devkit/build-angular:tslint',
              options: {
                tsConfig: [
                  'libs/lib1/tsconfig.lib.json',
                  'libs/lib1/tsconfig.spec.json',
                ],
                exclude: ['**/node_modules/**', '!libs/lib1/**/*'],
              },
            },
          },
        });
        workspace.projects.add({
          name: 'app2',
          root: 'apps/app2',
          sourceRoot: 'apps/app2/src',
          projectType: 'application',
          prefix: 'app2',
          architect: {
            build: {
              builder: '@nrwl/node:build',
              outputs: ['{options.outputPath}'],
              options: {
                outputPath: 'dist/apps/app2',
                main: 'apps/app2/src/main.ts',
                tsConfig: 'apps/app2/tsconfig.app.json',
                assets: ['apps/app2/src/assets'],
              },
              configurations: {
                production: {
                  optimization: true,
                  extractLicenses: true,
                  inspect: false,
                  fileReplacements: [
                    {
                      replace: 'apps/app2/src/environments/environment.ts',
                      with: 'apps/app2/src/environments/environment.prod.ts',
                    },
                  ],
                },
              },
            },
            serve: {
              builder: '@nrwl/node:execute',
              options: {
                buildTarget: 'app2:build',
              },
            },
            lint: {
              builder: '@angular-devkit/build-angular:tslint',
              options: {
                tsConfig: [
                  'apps/app2/tsconfig.app.json',
                  'apps/app2/tsconfig.spec.json',
                ],
                exclude: ['**/node_modules/**', '!apps/app2/**/*'],
              },
            },
            test: {
              builder: '@nrwl/jest:jest',
              outputs: ['coverage/apps/app2'],
              options: {
                jestConfig: 'apps/app2/jest.config.js',
                passWithNoTests: true,
              },
            },
          },
        });
        workspace.projects.add({
          name: 'app2-e2e',
          root: 'apps/app2-e2e',
          sourceRoot: 'apps/app2-e2e/src',
          projectType: 'application',
          architect: {
            e2e: {
              builder: '@angular-devkit/build-angular:protractor',
              options: {
                protractorConfig: 'apps/app2-e2e/protractor.conf.js',
                devServerTarget: 'app2:serve',
              },
              configurations: {
                production: {
                  devServerTarget: 'app2:serve:production',
                },
              },
            },
            lint: {
              builder: '@angular-devkit/build-angular:tslint',
              options: {
                tsConfig: ['apps/app2-e2e/tsconfig.e2e.json'],
                exclude: ['**/node_modules/**', '!apps/app2-e2e/**/*'],
              },
            },
          },
        });
      }),
      tree
    );
    tree.create('apps/app2-e2e/protractor.conf.js', ``); // will be used to determine that it is a Protractor E2E project
    tree.create('libs/lib1/karma.conf.js', ``); // will be used to determine that it is an Angular library
    tree = await callRule(
      chain([
        updateJsonInTree('tslint.json', () => exampleRootTslintJson.raw),
        updateJsonInTree(
          'apps/app1/tslint.json',
          () => exampleAngularProjectTslintJson.raw
        ),
        updateJsonInTree(
          'apps/app1-e2e/tslint.json',
          () => exampleE2eProjectTslintJson.raw
        ),
        updateJsonInTree(
          'libs/lib1/tslint.json',
          () => exampleAngularProjectTslintJson.raw
        ),
        updateJsonInTree(
          'apps/app2/tslint.json',
          () => exampleNonAngularProjectTslintJson.raw
        ),
      ]),
      tree
    );
    // Source files used to verify the comment conversion logic
    tree.create(
      'libs/lib1/foo.ts',
      `
// tslint:disable-next-line:no-eval
eval('');
    `.trim()
    );
    tree.create(
      'libs/lib1/bar.ts',
      `
/* tslint:disable */
eval('');
/* tslint:enable */
    `.trim()
    );
  });

  it('should update the lint target for the given project in the angular.json', async () => {
    async function runSchematicForProjectAndReadProjectConfig(
      tree: Tree,
      project: string
    ) {
      const updatedWorkspace = await readWorkspace(
        await runSchematic(
          'convert-tslint-to-eslint',
          <ConvertTSLintToESLintSchema>{
            project,
          },
          tree
        )
      );
      return updatedWorkspace.projects[project];
    }

    expect(
      await runSchematicForProjectAndReadProjectConfig(tree, 'app1')
    ).toMatchSnapshot();

    // Cypress E2E Project
    expect(
      await runSchematicForProjectAndReadProjectConfig(tree, 'app1-e2e')
    ).toMatchSnapshot();

    expect(
      await runSchematicForProjectAndReadProjectConfig(tree, 'lib1')
    ).toMatchSnapshot();

    expect(
      await runSchematicForProjectAndReadProjectConfig(tree, 'app2')
    ).toMatchSnapshot();

    // Protractor E2E Project
    expect(
      await runSchematicForProjectAndReadProjectConfig(tree, 'app2-e2e')
    ).toMatchSnapshot();
  });

  it('should create a .eslintrc.json and delete the tslint.json for the given project', async () => {
    async function runSchematicForProjectAndReadESLintConfig(
      tree: Tree,
      project: string
    ) {
      const result = await runSchematic(
        'convert-tslint-to-eslint',
        <ConvertTSLintToESLintSchema>{
          project,
        },
        tree
      );
      const workspace = await readWorkspace(result);
      return readJsonInTree(
        result,
        join(normalize(workspace.projects[project].root), '.eslintrc.json')
      );
    }

    // All files - Before
    expect(allFilesInDirInHost(tree, tree.root.path)).toMatchSnapshot();

    expect(
      await runSchematicForProjectAndReadESLintConfig(tree, 'app1')
    ).toMatchSnapshot();

    // Root ESLint config (which should exist after the schematic has been run at least once)
    expect(await readJsonInTree(tree, '.eslintrc.json')).toMatchSnapshot();

    expect(
      await runSchematicForProjectAndReadESLintConfig(tree, 'app1-e2e')
    ).toMatchSnapshot();

    expect(
      await runSchematicForProjectAndReadESLintConfig(tree, 'lib1')
    ).toMatchSnapshot();

    expect(
      await runSchematicForProjectAndReadESLintConfig(tree, 'app2')
    ).toMatchSnapshot();

    // All files - After
    expect(allFilesInDirInHost(tree, tree.root.path)).toMatchSnapshot();
  });

  it('should add the relevant dependencies to the package.json', async () => {
    // Before
    expect(readJsonInTree(tree, 'package.json')).toMatchInlineSnapshot(`
      Object {
        "dependencies": Object {},
        "devDependencies": Object {},
        "name": "test-name",
      }
    `);

    await readWorkspace(
      await runSchematic(
        'convert-tslint-to-eslint',
        <ConvertTSLintToESLintSchema>{
          project: 'app1',
        },
        tree
      )
    );

    // After
    expect(readJsonInTree(tree, 'package.json')).toMatchInlineSnapshot(`
      Object {
        "dependencies": Object {},
        "devDependencies": Object {
          "@angular-eslint/eslint-plugin": "~1.0.0",
          "@angular-eslint/eslint-plugin-template": "~1.0.0",
          "@angular-eslint/template-parser": "~1.0.0",
          "@nrwl/eslint-plugin-nx": "*",
          "@typescript-eslint/eslint-plugin": "4.3.0",
          "@typescript-eslint/parser": "4.3.0",
          "eslint": "7.10.0",
          "eslint-config-prettier": "6.0.0",
          "eslint-plugin-import": "latest",
        },
        "name": "test-name",
      }
    `);
  });

  it('should replace tslint:disable comments with their eslint equivalents', async () => {
    // Before
    expect(tree.read('libs/lib1/foo.ts').toString()).toMatchInlineSnapshot(`
"// tslint:disable-next-line:no-eval
eval('');"
`);
    expect(tree.read('libs/lib1/bar.ts').toString()).toMatchInlineSnapshot(`
"/* tslint:disable */
eval('');
/* tslint:enable */"
`);

    await readWorkspace(
      await runSchematic(
        'convert-tslint-to-eslint',
        <ConvertTSLintToESLintSchema>{
          project: 'lib1',
        },
        tree
      )
    );

    // After
    expect(tree.read('libs/lib1/foo.ts').toString()).toMatchInlineSnapshot(`
"// eslint-disable-next-line no-eval
eval('');
"
`);
    expect(tree.read('libs/lib1/bar.ts').toString()).toMatchInlineSnapshot(`
"/* eslint-disable */
eval('');
/* eslint-enable */
"
`);
  });
});
