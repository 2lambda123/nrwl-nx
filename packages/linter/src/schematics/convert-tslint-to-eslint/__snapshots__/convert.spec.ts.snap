// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convert-tslint-to-eslint should create a .eslintrc.json and delete the tslint.json for the given project 1`] = `
Array [
  "/workspace.json",
  "/package.json",
  "/nx.json",
  "/tsconfig.base.json",
  "/tslint.json",
  "/apps/app2-e2e/protractor.conf.js",
  "/apps/app1/tslint.json",
  "/apps/app1-e2e/tslint.json",
  "/apps/app2/tslint.json",
  "/libs/lib1/karma.conf.js",
  "/libs/lib1/tslint.json",
  "/libs/lib1/foo.ts",
  "/libs/lib1/bar.ts",
]
`;

exports[`convert-tslint-to-eslint should create a .eslintrc.json and delete the tslint.json for the given project 2`] = `
Object {
  "extends": "../../.eslintrc.json",
  "ignorePatterns": Array [
    "!**/*",
  ],
  "overrides": Array [
    Object {
      "extends": Array [
        "plugin:@nrwl/nx/angular",
        "plugin:@angular-eslint/template/process-inline-templates",
      ],
      "files": Array [
        "*.ts",
      ],
      "parserOptions": Object {
        "project": Array [
          "apps/app1/tsconfig.*?.json",
        ],
      },
      "rules": Object {
        "@angular-eslint/component-selector": Array [
          "error",
          Object {
            "prefix": "foo",
            "style": "kebab-case",
            "type": "element",
          },
        ],
        "@angular-eslint/directive-selector": Array [
          "error",
          Object {
            "prefix": "foo",
            "style": "camelCase",
            "type": "attribute",
          },
        ],
      },
    },
    Object {
      "extends": Array [
        "plugin:@nrwl/nx/angular-template",
      ],
      "files": Array [
        "*.html",
      ],
      "rules": Object {},
    },
  ],
}
`;

exports[`convert-tslint-to-eslint should create a .eslintrc.json and delete the tslint.json for the given project 3`] = `
Object {
  "ignorePatterns": Array [
    "**/*",
  ],
  "overrides": Array [
    Object {
      "files": Array [
        "*.ts",
        "*.tsx",
        "*.js",
        "*.jsx",
      ],
      "rules": Object {
        "@nrwl/nx/enforce-module-boundaries": Array [
          "error",
          Object {
            "allow": Array [
              "@nx-example/shared/product/data/testing",
            ],
            "depConstraints": Array [
              Object {
                "onlyDependOnLibsWithTags": Array [
                  "type:feature",
                  "type:ui",
                ],
                "sourceTag": "type:app",
              },
              Object {
                "onlyDependOnLibsWithTags": Array [
                  "type:ui",
                  "type:data",
                  "type:types",
                  "type:state",
                ],
                "sourceTag": "type:feature",
              },
              Object {
                "onlyDependOnLibsWithTags": Array [
                  "type:types",
                ],
                "sourceTag": "type:types",
              },
              Object {
                "onlyDependOnLibsWithTags": Array [
                  "type:state",
                  "type:types",
                  "type:data",
                ],
                "sourceTag": "type:state",
              },
              Object {
                "onlyDependOnLibsWithTags": Array [
                  "type:types",
                ],
                "sourceTag": "type:data",
              },
              Object {
                "onlyDependOnLibsWithTags": Array [
                  "type:e2e-utils",
                ],
                "sourceTag": "type:e2e",
              },
              Object {
                "onlyDependOnLibsWithTags": Array [
                  "type:types",
                  "type:ui",
                ],
                "sourceTag": "type:ui",
              },
              Object {
                "onlyDependOnLibsWithTags": Array [
                  "scope:products",
                  "scope:shared",
                ],
                "sourceTag": "scope:products",
              },
              Object {
                "onlyDependOnLibsWithTags": Array [
                  "scope:cart",
                  "scope:shared",
                ],
                "sourceTag": "scope:cart",
              },
            ],
            "enforceBuildableLibDependency": true,
          },
        ],
      },
    },
    Object {
      "extends": Array [
        "plugin:@nrwl/nx/typescript",
      ],
      "files": Array [
        "*.ts",
        "*.tsx",
      ],
      "parserOptions": Object {
        "project": "./tsconfig.*?.json",
      },
      "rules": Object {},
    },
    Object {
      "extends": Array [
        "plugin:@nrwl/nx/javascript",
      ],
      "files": Array [
        "*.js",
        "*.jsx",
      ],
      "rules": Object {},
    },
  ],
  "plugins": Array [
    "@nrwl/nx",
  ],
  "root": true,
}
`;

exports[`convert-tslint-to-eslint should create a .eslintrc.json and delete the tslint.json for the given project 4`] = `
Object {
  "extends": Array [
    "plugin:cypress/recommended",
    "../../.eslintrc.json",
  ],
  "ignorePatterns": Array [
    "!**/*",
  ],
  "overrides": Array [
    Object {
      "files": Array [
        "src/plugins/index.js",
      ],
      "rules": Object {
        "@typescript-eslint/no-var-requires": "off",
        "no-undef": "off",
      },
    },
    Object {
      "files": Array [
        "*.ts",
      ],
      "parserOptions": Object {
        "project": Array [
          "apps/app1-e2e/tsconfig.*?.json",
        ],
      },
      "rules": Object {
        "@typescript-eslint/no-namespace": "off",
        "@typescript-eslint/no-unused-vars": "off",
      },
    },
  ],
}
`;

exports[`convert-tslint-to-eslint should create a .eslintrc.json and delete the tslint.json for the given project 5`] = `
Object {
  "extends": "../../.eslintrc.json",
  "ignorePatterns": Array [
    "!**/*",
  ],
  "overrides": Array [
    Object {
      "extends": Array [
        "plugin:@nrwl/nx/angular",
        "plugin:@angular-eslint/template/process-inline-templates",
      ],
      "files": Array [
        "*.ts",
      ],
      "parserOptions": Object {
        "project": Array [
          "libs/lib1/tsconfig.*?.json",
        ],
      },
      "rules": Object {
        "@angular-eslint/component-selector": Array [
          "error",
          Object {
            "prefix": "foo",
            "style": "kebab-case",
            "type": "element",
          },
        ],
        "@angular-eslint/directive-selector": Array [
          "error",
          Object {
            "prefix": "foo",
            "style": "camelCase",
            "type": "attribute",
          },
        ],
      },
    },
    Object {
      "extends": Array [
        "plugin:@nrwl/nx/angular-template",
      ],
      "files": Array [
        "*.html",
      ],
      "rules": Object {},
    },
  ],
}
`;

exports[`convert-tslint-to-eslint should create a .eslintrc.json and delete the tslint.json for the given project 6`] = `
Object {
  "extends": "../../.eslintrc.json",
  "ignorePatterns": Array [
    "!**/*",
  ],
  "overrides": Array [
    Object {
      "files": Array [
        "*.ts",
      ],
      "parserOptions": Object {
        "project": Array [
          "apps/app2/tsconfig.*?.json",
        ],
      },
      "rules": Object {},
    },
  ],
}
`;

exports[`convert-tslint-to-eslint should create a .eslintrc.json and delete the tslint.json for the given project 7`] = `
Array [
  "/workspace.json",
  "/package.json",
  "/nx.json",
  "/tsconfig.base.json",
  "/tslint.json",
  "/.eslintrc.json",
  "/apps/app2-e2e/protractor.conf.js",
  "/apps/app1/.eslintrc.json",
  "/apps/app1-e2e/.eslintrc.json",
  "/apps/app2/.eslintrc.json",
  "/libs/lib1/karma.conf.js",
  "/libs/lib1/foo.ts",
  "/libs/lib1/bar.ts",
  "/libs/lib1/.eslintrc.json",
]
`;

exports[`convert-tslint-to-eslint should update the lint target for the given project in the angular.json 1`] = `
Object {
  "architect": Object {
    "lint": Object {
      "builder": "@nrwl/linter:eslint",
      "options": Object {
        "lintFilePatterns": Array [
          "apps/app1/src/**/*.ts",
          "apps/app1/src/**/*.html",
        ],
      },
    },
    "serve": Object {
      "builder": "@angular-devkit/build-angular:dev-server",
      "configurations": Object {
        "production": Object {
          "browserTarget": "app1:build:production",
        },
      },
      "options": Object {
        "browserTarget": "app1:build",
      },
    },
  },
  "projectType": "application",
  "root": "apps/app1",
  "sourceRoot": "apps/app1/src",
}
`;

exports[`convert-tslint-to-eslint should update the lint target for the given project in the angular.json 2`] = `
Object {
  "architect": Object {
    "e2e": Object {
      "builder": "@nrwl/cypress:cypress",
      "configurations": Object {
        "production": Object {
          "devServerTarget": "app1:serve:production",
        },
      },
      "options": Object {
        "cypressConfig": "apps/app1-e2e/cypress.json",
        "devServerTarget": "app1:serve",
        "tsConfig": "apps/app1-e2e/tsconfig.e2e.json",
      },
    },
    "lint": Object {
      "builder": "@nrwl/linter:eslint",
      "options": Object {
        "lintFilePatterns": Array [
          "apps/app1-e2e/**/*.ts",
        ],
      },
    },
  },
  "projectType": "application",
  "root": "apps/app1-e2e",
  "sourceRoot": "apps/app1-e2e/src",
}
`;

exports[`convert-tslint-to-eslint should update the lint target for the given project in the angular.json 3`] = `
Object {
  "architect": Object {
    "lint": Object {
      "builder": "@nrwl/linter:eslint",
      "options": Object {
        "lintFilePatterns": Array [
          "libs/lib1/src/**/*.ts",
          "libs/lib1/src/**/*.html",
        ],
      },
    },
  },
  "projectType": "library",
  "root": "libs/lib1",
  "sourceRoot": "libs/lib1/src",
}
`;

exports[`convert-tslint-to-eslint should update the lint target for the given project in the angular.json 4`] = `
Object {
  "architect": Object {
    "build": Object {
      "builder": "@nrwl/node:build",
      "configurations": Object {
        "production": Object {
          "extractLicenses": true,
          "fileReplacements": Array [
            Object {
              "replace": "apps/app2/src/environments/environment.ts",
              "with": "apps/app2/src/environments/environment.prod.ts",
            },
          ],
          "inspect": false,
          "optimization": true,
        },
      },
      "options": Object {
        "assets": Array [
          "apps/app2/src/assets",
        ],
        "main": "apps/app2/src/main.ts",
        "outputPath": "dist/apps/app2",
        "tsConfig": "apps/app2/tsconfig.app.json",
      },
      "outputs": Array [
        "{options.outputPath}",
      ],
    },
    "lint": Object {
      "builder": "@nrwl/linter:eslint",
      "options": Object {
        "lintFilePatterns": Array [
          "apps/app2/**/*.ts",
        ],
      },
    },
    "serve": Object {
      "builder": "@nrwl/node:execute",
      "options": Object {
        "buildTarget": "app2:build",
      },
    },
    "test": Object {
      "builder": "@nrwl/jest:jest",
      "options": Object {
        "jestConfig": "apps/app2/jest.config.js",
        "passWithNoTests": true,
      },
      "outputs": Array [
        "coverage/apps/app2",
      ],
    },
  },
  "prefix": "app2",
  "projectType": "application",
  "root": "apps/app2",
  "sourceRoot": "apps/app2/src",
}
`;

exports[`convert-tslint-to-eslint should update the lint target for the given project in the angular.json 5`] = `
Object {
  "architect": Object {
    "e2e": Object {
      "builder": "@angular-devkit/build-angular:protractor",
      "configurations": Object {
        "production": Object {
          "devServerTarget": "app2:serve:production",
        },
      },
      "options": Object {
        "devServerTarget": "app2:serve",
        "protractorConfig": "apps/app2-e2e/protractor.conf.js",
      },
    },
    "lint": Object {
      "builder": "@nrwl/linter:eslint",
      "options": Object {
        "lintFilePatterns": Array [
          "apps/app2-e2e/**/*.ts",
        ],
      },
    },
  },
  "projectType": "application",
  "root": "apps/app2-e2e",
  "sourceRoot": "apps/app2-e2e/src",
}
`;
